src/sedpack/io/dataset_iteration.py

```diff
@@ -93,6 +100,19 @@ def shard_paths_dataset(
         if shards:
             shards_list = shards_list[:shards]
 
+        # Only use a limited amount of shards for each setting of
+        # custom_metadata.
+        if custom_metadata_type_limit:
+            counts: dict[tuple[tuple[str, Any], ...], int] = {}
+            old_shards_list = shards_list
+            shards_list = []
+            for shard_info in old_shards_list:
+                k = tuple(sorted(shard_info.custom_metadata.items()))
+                counts[k] = counts.get(k, 0) + 1
```

Comment:
Consider using `defaultdict`:
```
counts = defaultdict(int)
[...]
for shard_info in old_shards_list:
    k = ...
    counts[k] += 1
```

```

Reply:
Personally I prefer `counts.get(k, 0)` since by looking at a single line I can tell this is fine. With the defaultdict I need to remember the type of `counts`. Happy to change that if you find that more readable.

Reply:
nah, just a personal preference :)